agent/lib_aws_sdk_php.c

```diff
@@ -0,0 +1,166 @@
+/*
+ * Copyright 2024 New Relic Corporation. All rights reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/*
+ * Functions relating to instrumenting the AWS-SDK-PHP.
+ * https://github.com/aws/aws-sdk-php
+ */
+#include "php_agent.h"
+#include "php_wrapper.h"
+#include "fw_hooks.h"
+#include "fw_support.h"
+#include "util_logging.h"
+#include "php_call.h"
+#include "lib_aws_sdk_php.h"
+
+#define PHP_PACKAGE_NAME "aws/aws-sdk-php"
+#define PHP_AWS_CLASS_PREFIX "Supportability/library/aws/aws-sdk-php/"
+
+/*
+ * In a normal course of events, the following line will always work
+ * zend_eval_string("Aws\\Sdk::VERSION;", &retval, "Get AWS Version")
+ * By the time we have detected the existence of the aws-sdk-php and with
+ * default composer profject settings, it callable even from
+ * nr_aws_sdk_php_enable which will automatically load the class if it isn't
+ * loaded yet and then evaluate the string. However, in the rare case that files
+ * are not loaded via autoloader and/or have non-default composer classload
+ * settings, if the class is not found, PHP 8.2+ will generate a fatal
+ * unrecoverable uncatchable error error whenever it cannot find a class. While
+ * calling this from nr_aws_sdk_php_enable would have been great and would allow
+ * the sdk version value to be set only once, to avoid the very unlikely but not
+ * impossible fatal error, this will be called from the
+ * "Aws\\ClientResolver::_apply_user_agent" wrapper which GUARANTEES that
+ * aws/sdk exists and is already loaded.
+ *
+ *
+ * Additionally given that aws-sdk-php is currently detected from the
+ * AwsClient.php file, this method will always be called when a client is
+ * created. Having separate functionality to extract from Aws/Sdk::__constructor
+ * is both not required and is redundant and causes additional overhead.
+ */
+extern void lib_aws_sdk_php_handle_version() {
+  char* version = NULL;
+  zval retval;
+  int result = FAILURE;
+
+  result = zend_eval_string("Aws\\Sdk::VERSION;", &retval, "Get AWS Version");
```

Comment:
Can't we have some guard code that verifies the class exists before referencing it using `class_exists()`?

Reply:
We could, which would slightly add to the overhead, but the advantage of calling it from ClientResolver::_apply_user_agent is that it is guaranteed to already exist so it saves slightly on overhead.
We calculate this value in func end, so you see the [wrapped function ](https://new-relic.atlassian.net/browse/NR-287951)has already made the call to sdk::version, so if it didn't exist and were to fail, it would do so before it got to us.

Reply:
I guess as defense against an older or newer version of the sdk where those assumptions dont hold.  Otherwise there will be a surprise for customers.

Reply:
The older version (2.x) of the sdk is from 2016, so we won't support that anyway.
This function is dependable and will always get that info (it's what they use to build their headers for communication) so is likely to always be guaranteed to retrieve that information.

It does save us in overhead, but if you feel strongly that it outweighs the overhead gain I can add it.

Reply:
To future/past proof as well as harden against this being used to wrap against any other function besides the special one with assumed guarantees, the version extraction is reverted to [original](https://github.com/newrelic/newrelic-php-agent/pull/932/commits/2c27a8cd99e658414b3d5281c3322154d544021a#diff-b5410abe33a951704c8600c7b0b214d3c431cee94dcb3089567ec0ec9f328fc1R21-R37).

Please see [10d00a4](https://github.com/newrelic/newrelic-php-agent/pull/932/commits/10d00a4fbc4d339f55d36224a731cebbe3737272)